src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java

```diff
@@ -0,0 +1,487 @@
+     */
+    private static final ParseField ENDPOINT_FIELD = new ParseField("endpoint");
+    private static final ParseField STATE_FIELD = new ParseField("state");
+    private static final ParseField INDICES_FIELD = new ParseField("indices");
+    private static final ParseField DATABASE_FIELD = new ParseField("database");
+    private static final ParseField UPDATE_STATS_FIELD = new ParseField("update_stats");
+
+    /**
+     * Default variables for job scheduling
+     */
+
+    /**
+     * @param id Id of a datasource
+     * @return Id of a datasource
+     */
+    private String id;
+    /**
+     * @param lastUpdateTime Last update time of a datasource
+     * @return Last update time of a datasource
+     */
+    private Instant lastUpdateTime;
+    /**
+     * @param enabledTime Last time when a scheduling is enabled for a GeoIP data update
+     * @return Last time when a scheduling is enabled for the job scheduler
+     */
+    private Instant enabledTime;
+    /**
+     * @param isEnabled Indicate if GeoIP data update is scheduled or not
+     * @return Indicate if scheduling is enabled or not
+     */
+    private boolean isEnabled;
+    /**
+     * @param schedule Schedule for a GeoIP data update
+     * @return Schedule for the job scheduler
+     */
+    private Schedule schedule;
+
+    /**
+     * Additional variables for datasource
+     */
+
+    /**
+     * @param endpoint URL of a manifest file
+     * @return URL of a manifest file
+     */
+    private String endpoint;
+    /**
+     * @param state State of a datasource
+     * @return State of a datasource
+     */
+    private DatasourceState state;
+    /**
+     * @param indices A list of indices having GeoIP data
+     * @return A list of indices having GeoIP data
+     */
+    private List<String> indices;
+    /**
+     * @param database GeoIP database information
+     * @return GeoIP database information
+     */
+    private Database database;
+    /**
+     * @param updateStats GeoIP database update statistics
+     * @return GeoIP database update statistics
+     */
+    private UpdateStats updateStats;
+
+    /**
+     * Datasource parser
+     */
+    public static final ConstructingObjectParser<Datasource, Void> PARSER = new ConstructingObjectParser<>(
+        "datasource_metadata",
+        true,
+        args -> {
+            String id = (String) args[0];
+            Instant lastUpdateTime = Instant.ofEpochMilli((long) args[1]);
+            Instant enabledTime = args[2] == null ? null : Instant.ofEpochMilli((long) args[2]);
+            boolean isEnabled = (boolean) args[3];
+            Schedule schedule = (Schedule) args[4];
+            String endpoint = (String) args[5];
+            DatasourceState state = DatasourceState.valueOf((String) args[6]);
+            List<String> indices = (List<String>) args[7];
+            Database database = (Database) args[8];
+            UpdateStats updateStats = (UpdateStats) args[9];
```

Comment:
why we have these magic indices for getting a value? Is this the only way? this has a big potential of creating errors if we want to add or remove some arguments. Try to simplify this, or I would like to understand from where this order is getting decided.

Reply:
Order is decided in
```
    static {
        PARSER.declareString(ConstructingObjectParser.constructorArg(), ID_FIELD);
        PARSER.declareLong(ConstructingObjectParser.constructorArg(), LAST_UPDATE_TIME_FIELD);
        PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), ENABLED_TIME_FILED);
        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), ENABLED_FILED);
        PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> ScheduleParser.parse(p), SCHEDULE_FIELD);
        PARSER.declareString(ConstructingObjectParser.constructorArg(), ENDPOINT_FIELD);
        PARSER.declareString(ConstructingObjectParser.constructorArg(), STATE_FIELD);
        PARSER.declareStringArray(ConstructingObjectParser.constructorArg(), INDICES_FIELD);
        PARSER.declareObject(ConstructingObjectParser.constructorArg(), Database.PARSER, DATABASE_FIELD);
        PARSER.declareObject(ConstructingObjectParser.constructorArg(), UpdateStats.PARSER, UPDATE_STATS_FIELD);

    }
```

Reply:
Can we do something better here? The problem is it will cause issues in future if we want to add or remove fields

Reply:
It won't break when we add a new field. 

Removing field is risky whichever approach we takes.

Reply:
is there a better a way to get the values? rather than getting it via index of an array.

Reply:
Discussed in offline.

Reply:
Discussed offline. No better way. Documentation can be found in ConstructingObjectParser class.